### PR TITLE
Blog posts/using aws ec2 and ecs to host hundreds of services

### PR DESCRIPTION
**Pull request by steverabino**
## Using AWS EC2 and ECS to host hundreds of services

A re-write of [my Workplace tech post](https://deliveroo.workplace.com/notes/tech-org/demystifying-the-platform-aws-ec2-and-ecs/904804776646710/) that gives some basics into:
- What EC2 is, and what we use it for
- What an ECS Cluster is, and how we autoscale it
- How we update our EC2 Amazon Machine Image while maintaining autoscaling

### CODE REVIEW

File: _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+It comes preinstalled with the software required to make it a usable computer;
+like an operating system (you can just assume Linux for now - others are
+available though), so it can be booted up and can run processes - more on that
+later…
+
+## What do we use EC2 hosts for?
+
+A few different uses, but the most common use is in an
+[ECS Cluster](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/clusters.html),
+a grouping of EC2 machines used as a home for ECS tasks - these are the
```
Comment (bencord0) [inline _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md:48]:
```suggestion
a grouping of EC2 machines used as a home for ECS Tasks - these are the
```

File: _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+for the most part, we leave it to ECS to decide.
+
+## What happens if the cluster is full?
+
+We are constantly monitoring the ECS cluster, and autoscale EC2 instances based
+on how much spare capacity there is. If there’s not enough spare capacity to
+immediately run another 40 large docker containers, we bump up the desired count
+of EC2 instances in the cluster, and EC2 spins up new machines (the number of
+machines we start up depends on how much below approximately 40 large container
+capacity we are). New EC2 instances can take up to 10 minutes before they’re
```
Comment (bencord0) [inline _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md:83]:
```suggestion
capacity we are). New EC2 instances can take a few minutes before they’re
```

I think our current timing is about two and a half minutes.

File: _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+ready to be used by ECS, so we need to have quite a large buffer to deal with
+dinnertime spikes in demand.
```
Comment (bencord0) [inline _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md:85]:
```suggestion
unexpected spikes in demand.
```

File: _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+## How do we change or upgrade the machine image used?
+
+Circling back to the software that is preinstalled on these EC2 servers. When
+booted up, an Amazon Machine Image (AMI) is used, which installs the basic
```
Comment (bencord0) [inline _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md:90]: The tools are baked into the AMI. On boot, the only required tasks are to register the speicifc hosts into the cluster, download some config and start running the Docker containers.

File: _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+tools you need to make the machine usable. Amazon provides a Linux AMI which we
```
Comment (bencord0) [inline _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md:91]: You can mention that we use packer and ansible to create our own Amazon Linux derived images.

File: _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+have built upon in the past to create our own machine image. We use this to run
+scripts which give all our running ECS tasks access to things that all
+Deliveroo’s services will need, such as software (like the
+[Datadog agent](https://docs.datadoghq.com/agent/) which sends metrics to
```
Comment (bencord0) [inline _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md:95]: Do you want to also mention other agents that we run? e.g. AWS Inspector, Cloudstrike Falcon and a few Deliveroo specific agents (e.g. for healthchecks)

File: _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+interface allows you to obtain and configure capacity with minimal friction. It
+provides you with complete control of your computing resources and lets you run
+on Amazon’s proven computing environment."
+
+_Amazon's description of EC2_
+</aside>
+
+I would say this; an EC2 host is a server. More simply, it is a computer. It is
+(in most cases) not an actual physical server in a rack, and Amazon abstracts
+that detail away us, but I find I get my head around the concept easier by
```
Comment (Crunch09) [inline _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md]:
```suggestion
that detail away for us, but I find I get my head around the concept easier by
```

File: _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+## What do we use EC2 hosts for?
+
+A few different uses, but the most common use is in an
+[ECS Cluster](https://docs.aws.amazon.com/AmazonECS/latest/developerguide/clusters.html),
+a grouping of EC2 machines used as a home for ECS Tasks - these are the
+Dockerized containers of our applications that are running with a command
+specified by the engineer in a config file.
+
+## ECS? what’s that, and how is it related to EC2?
```
Comment (Crunch09) [inline _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md]:
```suggestion
## ECS? What’s that, and how is it related to EC2?
```

File: _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+running ECS tasks access to things that all Deliveroo’s services will need, such
+as software (like the [Datadog agent](https://docs.datadoghq.com/agent/) which
+sends metrics to Datadog, and
+[AWS Inspector](https://aws.amazon.com/inspector/), AWS's automated security
+assessment service), roles, security policies, and environment variables that
+we need to apply to the containers.
+
+The process of rolling out a new machine image when an update is available, or
+when we make changes to our custom machine image, is not as straightforward as
+I’m used to (I have a new-found appreciation for release management software).
```
Comment (Crunch09) [inline _posts/2020-06-08-using-aws-ec2-and-ecs-to-host-hundreds-of-services.md]: Do you mean `as I would have expected` here or actually `as I'm used to` ? Just wondering if it's actually simpler with other providers?

**Pull request merged**